### PR TITLE
Log terraform provider warn and error in the right way

### PR DESCRIPTION
**Pull request by hemslo**
Previous changes reverted in https://github.com/app-sre/qontract-reconcile/pull/3707 because monkey patch can't work with already multithreaded environment.

In this change, `python-terraform` is removed and replaced by [lean_terraform_client](https://github.com/app-sre/qontract-reconcile/blob/2e931bfcbd4c4da8387c55edc8699b651f507380/reconcile/utils/lean_terraform_client.py), then reapply extract terraform provider logs logic from previous PRs:

* https://github.com/app-sre/qontract-reconcile/pull/3689
* https://github.com/app-sre/qontract-reconcile/pull/3705

[APPSRE-7884](https://issues.redhat.com/browse/APPSRE-7884)

Doc [Running Terraform in Automation](https://developer.hashicorp.com/terraform/tutorials/automation/automate-terraform)

### <samp>🤖 Generated by Copilot at 2e931bf</samp>

This pull request replaces the python_terraform package with the lean_tf module for the TerraformClient class, and adds unit tests and documentation for the terraform-related modules. It also removes the unused python_terraform dependency from the setup.py file.

### <samp>🤖 Generated by Copilot at 2e931bf</samp>

*  Refactor the TerraformClient class to use the lean_terraform_client module instead of the python_terraform package, and add a new dataclass TerraformSpec and a new exception class TerraformCommandError ([link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL14-R21), [link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL22-L26), [link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfR43-R49), [link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfR58-R67), [link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL84-R112), [link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL112-R155), [link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL132-R166), [link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL145-R177), [link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL171-R212), [link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL219-R255), [link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-5d6f3e1e250b0217592c7e87cd1048ac0f03e7f9ae07e979d540bb9170803bbfL231-R263))
*  Add type annotations, docstrings, and refactor the show_json function in the lean_terraform_client module, which provides a wrapper for the terraform commands using the subprocess module ([link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-f5b40651f6729aefce87470bb42a7bd4d4131e0955abd85a4c9e83f4f45b5b0fL3-R10), [link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-f5b40651f6729aefce87470bb42a7bd4d4131e0955abd85a4c9e83f4f45b5b0fL16-R140))
*  Remove the python_terraform package from the install_requires list in the setup.py file ([link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7L32))
*  Add unit tests for the check_output method of the TerraformClient class, which logs the output of the terraform commands based on the return code and the log level ([link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-b3ecbf5e85f99cd579b16e0340930eb7183e47033610cf8a42fa7cc71e45c54aR646-R906))
*  Add import statements, constants, and a pytest_mock fixture for the unit tests in the test_terraform_client.py file ([link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-b3ecbf5e85f99cd579b16e0340930eb7183e47033610cf8a42fa7cc71e45c54aR2-R3), [link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-b3ecbf5e85f99cd579b16e0340930eb7183e47033610cf8a42fa7cc71e45c54aR10), [link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-b3ecbf5e85f99cd579b16e0340930eb7183e47033610cf8a42fa7cc71e45c54aL22-R30))
*  Add import statements for the unit tests in the test_lean_terraform_client.py file ([link](https://github.com/app-sre/qontract-reconcile/pull/3708/files?diff=unified&w=0#diff-f1162b13b774db976e0022ff57e3c8cfa9406dab21e11a3e31ad21bf6431bb77R1-R181))

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -29,7 +29,6 @@
         "anymarkup>=0.7.0,<0.9.0",
         "python-gitlab>=1.11.0,<1.12.0",
         "semver~=2.13",
-        "python-terraform>=0.10.0,<0.11.0",
```
Comment (kwilczynski) [inline setup.py:32]: I love it!

**Comment by janboll** [comment]
Did you run the integration locally? init/plan/apply?

**Review by janboll — APPROVED** [review]
:100:

**Review by kwilczynski — APPROVED** [review]
@hemslo, looks good!

**Comment by hemslo** [comment]
> Did you run the integration locally? init/plan/apply?

@janboll yes, applied in dev data env, created a RDS instance to test.

new component test added in https://github.com/app-sre/qontract-reconcile/pull/3708/commits/2e931bfcbd4c4da8387c55edc8699b651f507380

**Pull request merged**